### PR TITLE
[Enhancement] Make segment iterator scan full chunk in each time 

### LINKED ISSUE #5699: make segment iterator scan full chunk in each time
## Enhancement
In each call to `_do_get_next`,  segment iterators will return a result chunk with all of its element passing the predicates _filter. To acheive this, we will continue to read more data(by `_read`) and filter them (by `_filter`), until we reach the chunk_capacity. However, when the chunk is nearly full, such as it already contains 4095 rows(capacity is 4096),  wo only fetch one more row in `_read`, and check it throught `_filter` function, when there are a number of rows unsatisfied the predicates, the scan speed will degrade a log.
```c++
  while ((chunk_start < chunk_capacity) & _range_iter.has_more()) {
      RETURN_IF_ERROR(_read(chunk, rowid, chunk_capacity - chunk start));
      chunk->check_or_die();
      size_t next_start = chunk->num_rows();

      if (has_predicate) {
          next_start = _filter(chunk, rowid, chunk_start, next_start);
          chunk->check_or_die();
      }
      chunk_start = next_start;
      DCHECK_EQ(chunk_start, chunk->num_rows());
  }
```

### PR DESCRIPTION
**Pull request by xiaoyong-z**
[Enhancement]: make segment iterator scan full chunk in each time 
Signed-off-by: xyz <a997647204@gmail.com>

## What type of PR is this：
- [ ] bug
- [ ] feature
- [x] enhancement
- [ ] others

## Which issues of this PR fixes ：

Fixes #5699

## Problem Summary(Required) ：

In each `_do_get_next`, segment iterators will return a result chunk with all of its elements passing the predicates _filter. To achieve this, it will scan some(chunk_capacity - chunk current size) rows, and filters these rows that are unsatisfied with the predicates. Assume the chunk size is 4096 when the chunk current size reaches 4095. Segment iterator will only scan one more row in `_read` and `_filter` it, when there are a number of rows unsatisfied with the predicates, this will have degradation to the performance.

To solve this problem, I make the segment iterator `_read` scan full rows(4096 rows, chunk_capacity) each time, and _filter 4096 rows in batch, which improves a lot to the scan speed. However, this may lead to the resulting chunk containing more than 4096 rows finally. We will save these overflow rows in _overflow_read_chunk chunk and reload them in next do_get_next.

This pr makes the scan speed has an overall improvement, in my test case, the best improvement is around 30%

### CODE REVIEW

**Comment by xiaoyong-z** [comment]
# Performance 
## Generate data script
```python
import random
import radar
from tqdm import tqdm
def tiny_generator():
    return str(random.randint(-128,127))

def smallint_generator():
    return str(random.randint(-32768, 32767))

def bigint_generator():
    return str(random.randint(-9223372036854775808, 9223372036854775807))

def int_generator():
    return str(random.randint(-2147483648, 2147483647))

def datetime_generator():
    return radar.random_datetime().strftime("%Y-%m-%d")

numbers = 500000000
with open("/data/xiaoyongzhao/output/8501/8501-4.csv",'w') as f2: 
    for i in tqdm(range(numbers)):
        key1 = int_generator()
        key2 = str(i % 10000000)
        key3 = str(i % 1000000)
        key4 = str(i % 100000)
        key5 = str(i % 10000)
        key6 = str(i % 1000)
        key7 = str(i % 100)
        key8 = str(i % 10)
        key9 = str(smallint_generator())
        key10 = str(i % 100000000)
        key11 = datetime_generator()
        key12 = datetime_generator()
        key13 = datetime_generator()
        value1 = bigint_generator()
        value2 = bigint_generator()
        value3 = bigint_generator()
        value4 = bigint_generator()
        update_time = datetime_generator()
        line = key1 + ',' + key2 + ',' + key3 + ',' + key4 + ',' + key5 + ',' + key6 + ',' + key7 + ',' + key8 + ',' + key9 + ',' + key10 + ',' + key11 + ',' + key12 + ',' + key13 + ',' + value1 + ',' + value2 + ',' + value3 + ',' + value4 + ',' + update_time
        f2.write(line + '\n')
f2.close()
```

## table
```
CREATE TABLE `table8501noindex` (
  `key1` int NOT NULL COMMENT "",
  `key2` varchar(20) NOT NULL COMMENT "",
  `key3` varchar(20) NOT NULL COMMENT "",
  `key4` varchar(20) NOT NULL COMMENT "",
  `key5` varchar(20) NOT NULL COMMENT "",
  `key6` varchar(20) NOT NULL COMMENT "",
  `key7` varchar(20) NOT NULL COMMENT "",
  `key8` varchar(20) NOT NULL COMMENT "",
  `key9` varchar(20) NOT NULL COMMENT "",
  `key10` varchar(20) NOT NULL DEFAULT "baishan" COMMENT "",
  `key11` datetime NOT NULL COMMENT "",
  `key12` datetime NOT NULL COMMENT "",
  `key13` datetime NOT NULL COMMENT "",
  `value1` bigint NULL COMMENT "",
  `value2` bigint NULL COMMENT "",
  `value3` bigint NULL COMMENT "",
  `value4` bigint NULL DEFAULT "0" COMMENT "",
  `update_time` datetime REPLACE NULL COMMENT ""
) ENGINE=OLAP
DUPLICATE KEY(`key1`)
COMMENT "OLAP"
DISTRIBUTED BY HASH(`key1`) BUCKETS 1
PROPERTIES (
"replication_num" = "1",
"in_memory" = "false",
"storage_format" = "V2"
);
```

Generate around 20 billion random data

**Comment by chaoyli** [comment]
Do you watch the memory consumption when the number of segments is large?
It will consume more memory.

**Comment by xiaoyong-z** [comment]
> Do you watch the memory consumption when the number of segments is large? It will consume more memory.

I reuse previous test, select count(*) from table where key2 == '1', it contains around 2840 segments. (haven't compaction)
Before optimization, the be memory is 204MB
After optimization, the be memory is 212MB

For 1GB segment, it mostly increase one chunk size memory, now chunk default size is 4096.

**Comment by xiaoyong-z** [comment]
After use update key, the previously mem starrocks_be_memory_allocated_bytes are 2.04G, after is 2.58G

**Review by chaoyli — DISMISSED** [review]

File: be/src/storage/rowset/segment_iterator.cpp [hunk 1]
```diff
@@ -682,6 +699,25 @@ Status SegmentIterator::_do_get_next(Chunk* result, vector<rowid_t>* rowid) {
...
+        }
+        DCHECK(_overflow_read_chunk->is_empty());
+        _overflow_read_chunk->append(*chunk, chunk_capacity, chunk_start - chunk_capacity);
+        if (rowid != nullptr) {
+            DCHECK(_overflow_read_chunk_rowids.empty());
+            _overflow_read_chunk_rowids.insert(_overflow_read_chunk_rowids.end(), rowid->begin() + chunk_capacity,
+                                               rowid->end());
+            rowid->resize(chunk_capacity);
+        }
+        chunk->set_num_rows(chunk_capacity);
```
Comment (meegoo) [inline be/src/storage/rowset/segment_iterator.cpp:718]: chunk->set_num_rows() `resize` data but not `shrink_to_fit`, it make all chunk actual memory consumption larger than 4096 rows?

**Comment by meegoo** [comment]
> After use update key, the previously mem starrocks_be_memory_allocated_bytes are 2.04G, after is 2.67G

It seems that memory usage is not a problem that can be ignored

File: be/src/storage/rowset/segment_iterator.cpp (hunk 1 above)
Comment (xiaoyong-z) [reply be/src/storage/rowset/segment_iterator.cpp:718]: Each time `do_get_next`, `chunk` is likely to contain more than chunk_capacity rows. I think `shrink_to_fit` here is not a good choice? If we do so, the frequent shrink and grow will have some overhead?

Comment (meegoo) [reply be/src/storage/rowset/segment_iterator.cpp:718]: I don't mean to call shrink_to_fit, but is this way of consuming extra memory worth it?

Comment (meegoo) [reply be/src/storage/rowset/segment_iterator.cpp:718]: previous implementation _read() size is `chunk_capacity - chunk_start`, Doesn't seem to exceed chunk capacity

**Comment by xiaoyong-z** [comment]
> > After use update key, the previously mem starrocks_be_memory_allocated_bytes are 2.04G, after is 2.67G
> 
> It seems that memory usage is not a problem that can be ignored

In this test, i load around 2 billion rows of data (with 2840 full segments) in one machine. Test it with update key(merge iterator), and no segment iterators can be early released.
I think this is the worst case, in most times, i think this modification does not consume that much memory.

**Comment by meegoo** [comment]
> > > After use update key, the previously mem starrocks_be_memory_allocated_bytes are 2.04G, after is 2.67G
> > 
> > 
> > It seems that memory usage is not a problem that can be ignored
> 
> In this test, i load around 2 billion rows of data (with 2840 full segments) in one machine. Test it with update key(merge iterator), and no segment iterators can be early released. I think this is the worst case, in most times, i think this modification does not consume that much memory.

We have a large number of versions(segments) of the scene, which can also lead to a large amount of memory

File: be/src/storage/rowset/segment_iterator.cpp (hunk 1 above)
Comment (xiaoyong-z) [reply be/src/storage/rowset/segment_iterator.cpp:718]: When chunk_start equals to `chunk_capacity - 1`, _read only scan one more line each time. If the row satisfied the predicate is last row in the worst case. Each time `_read` can only retrieve one line, the overhead of the function call will be very large.

**Comment by xiaoyong-z** [comment]
I limit the overflow chunk to 1/4 whole chunk. 

There is something wrong with the previous statistics. 
After use update key, 
before optimiztion;                        the mem starrocks_be_memory_allocated_bytes are 2.18G,
no limit to overflow chunk:           the previously mem starrocks_be_memory_allocated_bytes are 2.67G,
after limit the overflow chunk:     the previously mem starrocks_be_memory_allocated_bytes are 2.37G,

**Review by meegoo — APPROVED** [review]

**Review by chaoyli — APPROVED** [review]

**Review by Astralidea — APPROVED** [review]

File: be/src/storage/rowset/segment_iterator.cpp [hunk 2]
```diff
@@ -770,7 +806,7 @@ void SegmentIterator::_switch_context(ScanContext* to) {
     }
 
     if (to->_read_chunk == nullptr) {
-        to->_read_chunk = ChunkHelper::new_chunk(to->_read_schema, _opts.chunk_size);
+        to->_read_chunk = ChunkHelper::new_chunk(to->_read_schema, _opts.chunk_size + _opts.chunk_size / 4 + 1);
```
Comment (stdpain) [inline be/src/storage/rowset/segment_iterator.cpp:809]: It seems we could return a larger chunk to executor layer. which could cause Aggregator crash. So I want to revert this PR. @xiaoyong-z @chaoyli @trueeyu

Comment (stdpain) [reply be/src/storage/rowset/segment_iterator.cpp:809]: our Aggregator couldn't handle greater than `runtime_state->chunk_size()` Chunk

**Pull request merged**